modyn/supervisor/internal/triggers/datadrifttrigger.py

```diff
@@ -0,0 +1,333 @@
+
+    def inform(self, new_data: list[tuple[int, int, int]]) -> Generator[int]:
+        # logger.debug(f"[DataDriftDetector][Prev Trigger {self.previous_trigger_id}]"
+        #     + f"[Dataset {self.dataloader_info.dataset_id}]"
+        #     + f"[Cached data] {len(self.data_cache)}, [New data] {len(new_data)}")
+
+        # add new data to data_cache
+        self.data_cache.extend(new_data)
+
+        cached_data_points = len(self.data_cache)
+        total_data_for_detection = len(self.data_cache)
+        detection_data_idx_start = 0
+        detection_data_idx_end = 0
+        while total_data_for_detection >= self.detection_data_points:
+            total_data_for_detection -= self.detection_data_points
+            detection_data_idx_end += self.detection_data_points
+            if detection_data_idx_end <= self.untriggered_detection_data_points:
+                continue
+
+            # trigger id doesn't always start from 0, but always increments by 1
+            if self.previous_trigger_id is None:
+                # if no previous model exists, always trigger
+                triggered = True
+            else:
+                # if exist previous model, detect drift
+                triggered = self.detect_drift(detection_data_idx_start, detection_data_idx_end)
+
+            if triggered:
+                trigger_data_points = detection_data_idx_end - detection_data_idx_start
+                # logger.debug(f"[DataDriftDetector][Prev Trigger {self.previous_trigger_id}]"
+                #   + f" new{len(new_data)} cache{cached_data_points} trigger{trigger_data_points}")
+                trigger_idx = len(new_data) - (cached_data_points - trigger_data_points) - 1
+
+                logger.debug(
+                    f"[DataDriftDetector][Prev Trigger {self.previous_trigger_id}]"
+                    + f"[Dataset {self.dataloader_info.dataset_id}]"
+                    + f"[Trigger data points] {trigger_data_points}, [Trigger index] {trigger_idx}"
+                )
+                _, timestamps, _ = zip(
+                    *self.data_cache[detection_data_idx_start:detection_data_idx_end]
+                )  # type: ignore
+                num_per_t = {}
+                for t in timestamps:
+                    num_per_t[t] = num_per_t.get(t, 0) + 1
+                logger.debug(
+                    f"[DataDriftDetector][Prev Trigger {self.previous_trigger_id}]"
+                    + f"[Dataset {self.dataloader_info.dataset_id}]"
+                    + f"[Trigger Timestamps] {len(num_per_t)}:{num_per_t}"
+                )
+
+                # update bookkeeping and pointers
+                cached_data_points -= trigger_data_points
+                detection_data_idx_start = detection_data_idx_end
+                yield trigger_idx
+
+        # remove triggered data
+        self.data_cache = self.data_cache[detection_data_idx_start:]
```

Comment:
you can do `del self._data_cache[: detection_data_idx_start]` - this might induce less copies

Reply:
done